util/HalideTraceDump.cpp

```diff
@@ -0,0 +1,408 @@
+#include "HalideTraceUtils.h"
+#include "HalideBuffer.h"
+#include "halide_image_io.h"
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdint.h>
+#include <vector>
+#include <map>
+#include <string>
+#include <unistd.h>
+#include <fcntl.h>
+#include <string.h>
+
+/** \file
+ *
+ * A tool which can read a binary Halide trace file from stdin, and dump files containing the final pixel values
+ * recorded for each traced Func.
+ *
+ * Currently dumps into supported Halide image formats.
+ */
+
+using namespace Halide;
+using namespace Internal;
+
+using std::map;
+using std::vector;
+using std::string;
+using Halide::Runtime::Buffer;
+
+struct BufferOutputOpts {
+    enum OutputType {
+        PNG = 0,
+        JPG,
+        PGM
+    };
+
+    enum OutputType type;
+};
+
+int halide_copy_to_host(void *user_context, struct halide_buffer_t *buf) {
+    return 0;
+}
+
+struct Value {
+    uint8_t payload[8];
+    uint8_t defined[8];
+
+    Value() { defined[0] = 0; }
+    Value(Packet *p, int lane_idx = 0) {
+        defined[0] = 1;
```

Comment:
I wanted the array of Values to be directly wrappable in a Buffer. The Value struct's size is the payload (8 bytes, AFAIK the largest possible element of any Halide buffer) plus the "defined" property. I was playing it safe here, thinking that maybe a weird element size (like 9 bytes, or 8 + boolean size on whatever platform) would cause problems, so this way I made sure the total size would be a nice round 16 bytes.

If this rounding is unnecessary, or you have a better suggestion, I can make a change. For now I have changed it to a single "uint8_t defined" and a "uint8_t padding[7]", and added a comment to at least explain why this is here.